p2p/network.go

```diff
@@ -0,0 +1,152 @@
+package p2p
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/cmwaters/halo/consensus"
+	"github.com/cmwaters/halo/network"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var _ network.Network = (*Network)(nil)
+
+type Network struct {
+	ps *pubsub.PubSub
+}
+
+func NewNetwork(ps *pubsub.PubSub) network.Network {
+	return &Network{
+		ps: ps,
+	}
+}
+
+func (pn *Network) Gossip(namespace []byte) (network.Gossip, error) {
+	// TODO(@Wondertan): network ID prefix, so that we don't collide with random namespaces
+	//  that turn out to be the same
+	topic, err := pn.ps.Join(string(namespace))
```

Comment:
What's the tradeoff between having a single topic for both proposals and votes and splitting them into two different topics?

Reply:
Sharing my initiall thoughts regarding your question. It could be a bit messy to handle both proposals and votes in one handler:
* I guess message validation will not be the same for proposals and votes. Pubsub allows only 1 validateFn per topic, so in this case, validator func will be more complicated with additional `if else` statements for proposals and votes.
*  To identify where to unmarshal data, we will need additional field.
* IIRC there is a message queue in pubsub that consists of 32 messages, so when the buffer is full, then message can’t be delivered.